File: backend/src/main/java/com/workflow/api/controller/ProcessDefinitionController.java
package com.workflow.api.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.workflow.api.dto.PageResponse;
import com.workflow.api.dto.ProcessDefinitionSummary;
import com.workflow.common.domain.R;
import com.workflow.engine.process.ProcessService;
import com.workflow.engine.process.bpmn.InitiatorNodeResolver;
import com.workflow.engine.process.entity.NodeConfig;
import com.workflow.engine.process.entity.ProcessDraft;
import com.workflow.engine.process.repository.NodeConfigRepository;
import com.workflow.engine.process.repository.ProcessDraftRepository;
import org.flowable.engine.repository.ProcessDefinition;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/v1/deployed-processes")
public class ProcessDefinitionController {

    private final ProcessService processService;
    private final ProcessDraftRepository processDraftRepository;
    private final NodeConfigRepository nodeConfigRepository;
    private final InitiatorNodeResolver initiatorNodeResolver;
    private final ObjectMapper objectMapper;

    public ProcessDefinitionController(ProcessService processService,
                                       ProcessDraftRepository processDraftRepository,
                                       NodeConfigRepository nodeConfigRepository,
                                       InitiatorNodeResolver initiatorNodeResolver,
                                       ObjectMapper objectMapper) {
        this.processService = processService;
        this.processDraftRepository = processDraftRepository;
        this.nodeConfigRepository = nodeConfigRepository;
        this.initiatorNodeResolver = initiatorNodeResolver;
        this.objectMapper = objectMapper;
    }

    @GetMapping
    public R<PageResponse<Map<String, Object>>> list(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size,
            @RequestParam(required = false) String categoryId,
            @RequestParam(required = false) String name,
            @RequestParam(required = false) String status) {

        Page<ProcessDefinition> result = processService.listProcessDefinitions(
                PageRequest.of(page, size), categoryId, name, status);

        PageResponse<Map<String, Object>> response = new PageResponse<>(
                result.getContent().stream().map(this::toMap).toList(),
                result.getNumber(),
                result.getSize(),
                result.getTotalElements()
        );

        return R.ok(response);
    }

    /**
     * 已部署流程定义精简列表（按 key 去重取最新版本）。
     * 供调用活动子流程选择下拉使用。
     */
    @GetMapping("/summaries")
    public R<List<ProcessDefinitionSummary>> summaries() {
        return R.ok(processService.listSummaries());
    }

    @GetMapping("/{id}")
    public R<Map<String, Object>> get(@PathVariable String id) {
        return processService.getProcessDefinition(id)
                .map(def -> {
                    Map<String, Object> map = toMap(def);
                    // 查询流程关联的表单定义 ID
                    resolveFormDefIds(id, map);
                    return R.ok(map);
                })
                .orElse(R.fail(404, "Process definition not found"));
    }

    /**
     * 通过 NodeConfig 解析流程默认表单和发起人节点表单。
     * <p>优先级：发起人节点有表单配置则使用发起人节点表单，否则使用流程默认表单。
     * 结果写入 map：
     * <ul>
     *   <li>formDefId — 优先级最高的表单（发起人节点 > 流程默认）</li>
     *   <li>initiatorFormDefId — 发起人节点表单（单独返回，前端可区分）</li>
     *   <li>processFormDefId — 流程默认表单（单独返回，前端可区分）</li>
     * </ul>
     */
    private void resolveFormDefIds(String processDefinitionId, Map<String, Object> map) {
        if (processDefinitionId == null) {
            map.put("formDefId", null);
            map.put("initiatorFormDefId", null);
            map.put("processFormDefId", null);
            return;
        }
        // 精确匹配该部署版本的 NodeConfig 快照（部署时由当前配置复制生成）
        List<NodeConfig> configs = nodeConfigRepository.findByProcessDefinitionId(processDefinitionId);
        String processFormDefId = null;
        String initiatorFormDefId = null;

        for (NodeConfig nc : configs) {
            try {
                com.fasterxml.jackson.databind.JsonNode json = objectMapper.readTree(nc.getConfigJson());
                String formDefId = extractFormDefId(json);
                if (formDefId == null) continue;

                if ("__PROCESS__".equals(nc.getNodeId())) {
                    processFormDefId = formDefId;
                }
            } catch (Exception e) {
                // 忽略解析错误
            }
        }

        // 通过 InitiatorNodeResolver 获取发起人节点 ID，查其 NodeConfig 中的表单
        String initiatorNodeId = initiatorNodeResolver.resolve(processDefinitionId);
        if (initiatorNodeId != null) {
            for (NodeConfig nc : configs) {
                if (initiatorNodeId.equals(nc.getNodeId())) {
                    try {
                        com.fasterxml.jackson.databind.JsonNode json = objectMapper.readTree(nc.getConfigJson());
                        String formDefId = extractFormDefId(json);
                        if (formDefId != null) {
                            initiatorFormDefId = formDefId;
                        }
                    } catch (Exception e) {
                        // 忽略解析错误
                    }
                    break;
                }
            }
        }

        // 优先级：发起人节点表单 > 流程默认表单
        String effectiveFormDefId = initiatorFormDefId != null ? initiatorFormDefId : processFormDefId;

        map.put("formDefId", effectiveFormDefId);
        map.put("initiatorFormDefId", initiatorFormDefId);
        map.put("processFormDefId", processFormDefId);
    }

    /**
     * 从 NodeConfig JSON 中提取 formDefId。
     */
    private String extractFormDefId(com.fasterxml.jackson.databind.JsonNode json) {
        com.fasterxml.jackson.databind.JsonNode form = json.get("form");
        if (form != null && form.has("formDefId")) {
            String val = form.get("formDefId").asText();
            if (val != null && !val.isEmpty()) {
                return val;
            }
        }
        return null;
    }

    @GetMapping("/{id}/xml")
    public R<String> getXml(@PathVariable String id) {
        String xml = processService.getProcessDefinitionXml(id);
        return R.ok(xml);
    }

    @PostMapping("/{id}/suspend")
    public R<Void> suspend(@PathVariable String id) {
        processService.suspendProcessDefinition(id);
        return R.ok();
    }

    @PostMapping("/{id}/activate")
    public R<Void> activate(@PathVariable String id) {
        processService.activateProcessDefinition(id);
        return R.ok();
    }

    /**
     * 将 Flowable ProcessDefinition 安全序列化为 Map，
     * 避免直接 Jackson 序列化触发内部懒加载（processEngineConfiguration null 崩溃）。
     */
    private Map<String, Object> toMap(ProcessDefinition def) {
        Map<String, Object> map = new HashMap<>();
        map.put("id", def.getId());
        map.put("key", def.getKey());
        map.put("name", def.getName());
        map.put("version", def.getVersion());
        map.put("description", def.getDescription());
        map.put("deploymentId", def.getDeploymentId());
        map.put("resourceName", def.getResourceName());
        map.put("diagramResourceName", def.getDiagramResourceName());
        map.put("tenantId", def.getTenantId());
        map.put("category", def.getCategory());
        map.put("suspended", def.isSuspended());
        return map;
    }
}

File: backend-node/src/engine/process/controller/process-definition.controller.ts
import { Controller, Get, Param, Post, Query } from '@nestjs/common'
import { R } from '../../../common/domain/r'
import { PageResponse } from '../../../common/domain/page-response'
import { JavaStatusOk } from '../../../framework/http/java-status.decorator'
import { intQueryParam, assertPageSize } from '../../../framework/http/query-params'
import { getTenantId } from '../../../framework/tenant/tenant-context'
import {
  ProcessDesignService,
  type ProcessDefinitionSummaryVO,
  type ProcessVersionVO,
  type VersionEditorVO,
} from '../process-design.service'

/**
 * 已部署流程接口，对齐 Java `ProcessDefinitionController`
 * （前缀 `/api/v1/deployed-processes`）。
 *
 * 响应元素形状对齐 Flowable 的 `ProcessDefinition` 序列化（实测确认）：
 *   { id, key, name, version, deploymentId, resourceName, diagramResourceName,
 *     description, category, tenantId, suspended }
 * 其中 `id` 形如 `key:version:uuid`，`resourceName` 形如 `key.bpmn20.xml`，
 * `category` 取 BPMN 的 targetNamespace。
 */
@Controller('api/v1/deployed-processes')
@JavaStatusOk()
export class ProcessDefinitionController {
  constructor(private readonly service: ProcessDesignService) {}

  @Get()
  async list(
    @Query('page') page?: string,
    @Query('size') size?: string,
  ): Promise<R<PageResponse<Record<string, unknown>>>> {
    const tenantId = getTenantId()
    const safePage = Math.max(intQueryParam(page, 'page', 1), 1)
    const safeSize = intQueryParam(size, 'size', 20)
    // Java 用 `PageRequest.of(Math.max(page,1) - 1, size)` ⇒ size < 1 → HTTP 400
    assertPageSize(safeSize)

    const { rows, total } = await this.service
      .getRepository()
      .listProcessDefs(tenantId, (safePage - 1) * safeSize, safeSize)

    const content = rows.map((row) => {
      const key = String(row.process_key)
      const version = Number(row.version)
      return {
        id: String(row.id),
        key,
        name: row.name === null ? null : String(row.name),
        version,
        deploymentId: String(row.id),
        resourceName: `${key}.bpmn20.xml`,
        diagramResourceName: null,
        description: null,
        // Flowable 把 BPMN 的 targetNamespace 当作 ProcessDefinition.category
        category: row.target_namespace === null ? null : String(row.target_namespace),
        tenantId,
        suspended: String(row.status) !== 'ACTIVE',
      }
    })

    return R.ok(new PageResponse(content, safePage, safeSize, total))
  }

  /**
   * 已部署流程精简列表（供调用活动子流程选择下拉）。
   *
   * ⚠️ 必须声明在 `:id` **之前** —— 否则 `summaries` 会被当成流程定义 id。
   *    （Java 侧 `/summaries` 同样排在 `/{id}` 之前；Spring 另有「字面量优先」
   *    规则兜底，Nest 没有，只能靠声明顺序。）
   */
  @Get('summaries')
  async summaries(): Promise<R<ProcessDefinitionSummaryVO[]>> {
    return R.ok(await this.service.listSummaries())
  }

  /**
   * 某 key 的全部已部署版本（version 倒序，最高版本标记 latest）。
   *
   * 路径用**多段** `/key/{key}/versions`，与单段 `/{id}` 不会冲突（Java 亦如此）。
   */
  @Get('key/:key/versions')
  async versions(@Param('key') key: string): Promise<R<ProcessVersionVO[]>> {
    return R.ok(await this.service.listProcessVersions(key))
  }

  /**
   * 历史版本的编辑器数据（BPMN XML + 节点配置快照）。
   *
   * 读取失败 → `R.fail(404, '历史版本数据读取失败')`（HTTP 200 + body 内 code 404）。
   */
  @Get('versions/:procDefId/editor')
  async versionEditor(
    @Param('procDefId') procDefId: string,
  ): Promise<R<VersionEditorVO | null>> {
    const editor = await this.service.getVersionEditor(procDefId)
    if (editor === null) return R.fail(404, '历史版本数据读取失败')
    return R.ok(editor)
  }

  /**
   * 已部署流程定义详情。
   *
   * 不存在时 Java 返回 `R.fail(404, 'Process definition not found')` ——
   * HTTP 仍然是 **200**，404 只出现在 body 里。
   */
  @Get(':id')
  async get(@Param('id') id: string): Promise<R<Record<string, unknown> | null>> {
    const definition = await this.service.getProcessDefinition(id)
    if (definition === null) return R.fail(404, 'Process definition not found')
    return R.ok(definition)
  }

  /** 已部署流程的 BPMN XML（找不到时 data 为 null，状态仍是 200）。 */
  @Get(':id/xml')
  async xml(@Param('id') id: string): Promise<R<string | null>> {
    return R.ok(await this.service.getProcessDefinitionXml(id))
  }

  /** 挂起已部署流程。 */
  @Post(':id/suspend')
  async suspend(@Param('id') id: string): Promise<R<null>> {
    await this.service.suspendProcessDefinition(id)
    return R.ok()
  }

  /** 激活已部署流程。 */
  @Post(':id/activate')
  async activate(@Param('id') id: string): Promise<R<null>> {
    await this.service.activateProcessDefinition(id)
    return R.ok()
  }
}

